validator/lib/jsonapi/validator/resource.rb

```diff
@@ -0,0 +1,123 @@
+require 'jsonapi/validator/document'
+require 'jsonapi/validator/relationship'
+
+module JSONAPI
+  module Validator
+    class Resource
```

Comment:
I understand your concerns, and partly share them. Here are the motivations though:
1. Validating a JSONAPI document should be as fast as possible, hence instantiating a lot of objects (as `jsonapi-parser` does) is a problem;
2. Validating a JSONAPI document and providing an object-like API to traverse it are two different concerns that should be separated;
3. The JSONAPI spec seems to define "one format" but what it really defines is three: one for response documents, one for relationship creation/update payloads, and one for relationship update payloads. The latter two being somewhat narrower than the former, although the second allows absence of `id` for the primary data where the first does not.
4. This code is about validating the syntax of a document, and presence/absence of some attributes/relationships, as well as ensuring the value of some `type` members arguably falls within syntactic analysis, although this part is clearly domain-specific. This is the concern that I share with you, and I'd like to find an elegant solution to separate those.
5. I tend to find little value to the parser outside of validation. Maybe it could encapsulate more responsibility but I'm not sure what kind would be useful at this point.